File: tests/sample_go_project/calculator_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// PASSING TESTS
func TestAdd(t *testing.T) {
	assert.Equal(t, 5, Add(2, 3))
	assert.Equal(t, 0, Add(-1, 1))
	assert.Equal(t, 0, Add(0, 0))
}

func TestSubtract(t *testing.T) {
	assert.Equal(t, 2, Subtract(5, 3))
	assert.Equal(t, -5, Subtract(0, 5))
}

func TestMultiply(t *testing.T) {
	assert.Equal(t, 12, Multiply(3, 4))
	assert.Equal(t, 0, Multiply(0, 5))
}

// FAILING TEST - Intentional failure
func TestDivide(t *testing.T) {
	result, err := Divide(10, 2)
	assert.NoError(t, err)
	assert.Equal(t, 5.0, result)

	result, err = Divide(9, 3)
	assert.NoError(t, err)
	assert.Equal(t, 3.0, result)

	result, err = Divide(7, 2)
	assert.NoError(t, err)
	assert.Equal(t, 3.5, result)

	// FAILURE: Wrong assertion
	result, err = Divide(10, 4)
	assert.NoError(t, err)
	assert.Equal(t, 3.0, result) // Should be 2.5, this will FAIL
}

func TestDivideByZero(t *testing.T) {
	_, err := Divide(10, 0)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "cannot divide by zero")
}

// PARTIAL COVERAGE - ComplexCalculator only tested for some operations
func TestComplexCalculatorAdd(t *testing.T) {
	result, err := ComplexCalculator("add", 5, 3, nil)
	assert.NoError(t, err)
	assert.Equal(t, 8.0, result)
}

func TestComplexCalculatorMultiply(t *testing.T) {
	result, err := ComplexCalculator("multiply", 4, 5, nil)
	assert.NoError(t, err)
	assert.Equal(t, 20.0, result)
}

// FAILING TEST - Wrong expectation
func TestComplexCalculatorPower(t *testing.T) {
	result, err := ComplexCalculator("power", 2, 3, nil)
	assert.NoError(t, err)
	// FAILURE: Wrong expected value
	assert.Equal(t, 6.0, result) // Should be 8.0, this will FAIL
}

// NO TESTS for:
// - AdvancedMath() -> NO COVERAGE
// - HandleEdgeCases() -> NO COVERAGE
// - ComplexCalculator() other operations -> PARTIAL COVERAGE

File: tests/sample_javascript_project/src/calculator.test.ts
/**
 * Tests for calculator module.
 */

import {
  add,
  subtract,
  multiply,
  divide,
  complexCalculator
} from './calculator';

// PASSING TESTS
describe('Calculator', () => {
  describe('add', () => {
    it('should add two positive numbers', () => {
      expect(add(2, 3)).toBe(5);
    });

    it('should add negative and positive', () => {
      expect(add(-1, 1)).toBe(0);
    });

    it('should add zeros', () => {
      expect(add(0, 0)).toBe(0);
    });
  });

  describe('subtract', () => {
    it('should subtract numbers', () => {
      expect(subtract(5, 3)).toBe(2);
    });

    it('should handle negative results', () => {
      expect(subtract(0, 5)).toBe(-5);
    });
  });

  describe('multiply', () => {
    it('should multiply numbers', () => {
      expect(multiply(3, 4)).toBe(12);
    });

    it('should handle zero', () => {
      expect(multiply(0, 5)).toBe(0);
    });
  });

  // FAILING TEST - Intentional failure
  describe('divide', () => {
    it('should divide numbers', () => {
      expect(divide(10, 2)).toBe(5);
      expect(divide(9, 3)).toBe(3);
      expect(divide(7, 2)).toBe(3.5);
      // FAILURE: Wrong assertion
      expect(divide(10, 4)).toBe(3); // Should be 2.5, this will FAIL
    });

    it('should throw on division by zero', () => {
      expect(() => divide(10, 0)).toThrow('Cannot divide by zero');
    });
  });

  // PARTIAL COVERAGE - complexCalculator only tested for some operations
  describe('complexCalculator', () => {
    it('should handle add operation', () => {
      expect(complexCalculator('add', 5, 3)).toBe(8);
    });

    it('should handle multiply operation', () => {
      expect(complexCalculator('multiply', 4, 5)).toBe(20);
    });

    // FAILING TEST - Wrong expectation
    it('should handle power operation', () => {
      // FAILURE: Wrong expected value
      expect(complexCalculator('power', 2, 3)).toBe(6); // Should be 8, FAILS
    });
  });
});

// NO TESTS for:
// - advancedMath() -> NO COVERAGE
// - handleEdgeCases() -> NO COVERAGE
// - complexCalculator() other operations -> PARTIAL COVERAGE
